File: src/IECore/FileSequenceFunctions.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2009-2010, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include <algorithm>
#include <cassert>
#include <math.h>

#include "boost/version.hpp"
#include "boost/format.hpp"
#include "boost/lexical_cast.hpp"
#include "boost/regex.hpp"
#include "boost/filesystem/operations.hpp"
#include "boost/filesystem/path.hpp"
#include "boost/filesystem/convenience.hpp"
#include "boost/algorithm/string.hpp"

#include "IECore/Exception.h"
#include "IECore/FileSequence.h"
#include "IECore/FileSequenceFunctions.h"
#include "IECore/CompoundFrameList.h"
#include "IECore/EmptyFrameList.h"
#include "IECore/FrameRange.h"
#include "IECore/ReversedFrameList.h"

#if BOOST_VERSION < 103400

	// Boost versions prior to 1.34.0 performed unwanted file name checking. Disabling.
	static bool boostFilesystemDefaultNameCheckOverride()
	{
		boost::filesystem::path::default_name_check( boost::filesystem::no_check );
		return true;
	}
	static bool boostFilesystemDefaultNameCheckOverrideResult = boostFilesystemDefaultNameCheckOverride();

#endif

#if BOOST_VERSION < 104400

	// Boost 1.44.0 introduced Filesystem v3, which we use by defining BOOST_FILESYSTEM_VERSION=3 via
	// the build process. Prior versions of boost didn't have this version, so we need this define
	// to help write code suitable for both.

	#define PATH_TO_STRING filename()

#else

	#define PATH_TO_STRING filename().string()

#endif

using namespace IECore;

void IECore::findSequences( const std::vector< std::string > &names, std::vector< FileSequencePtr > &sequences, size_t minSequenceSize )
{
	sequences.clear();

	/// this matches names of the form $prefix$frameNumber$suffix
	/// placing each of those in a group of the resulting match.
	/// both $prefix and $suffix may be the empty string and $frameNumber
	/// may be preceded by a minus sign.
	/// It also matches file extensions with 3 or 4 characters that contain numbers (for example: CR2, MP3 )
	boost::regex matchExpression( std::string( "^([^#]*?)(-?[0-9]+)([^0-9#]*|[^0-9#]*\\.[a-zA-Z]{2,3}[0-9])$" ) );

	/// build a mapping from ($prefix, $suffix) to a list of $frameNumbers
	typedef std::vector< std::string > Frames;
	typedef std::map< std::pair< std::string, std::string >, Frames > SequenceMap;

	SequenceMap sequenceMap;

	for ( std::vector< std::string >::const_iterator it = names.begin(); it != names.end(); ++it )
	{
		boost::smatch matches;
		if ( boost::regex_match( *it, matches, matchExpression ) )
		{
			sequenceMap[
				SequenceMap::key_type(
					std::string( matches[1].first, matches[1].second ),
					std::string( matches[3].first, matches[3].second )
				)
			].push_back( std::string( matches[2].first, matches[2].second ) );
		}
	}

	for ( SequenceMap::const_iterator it = sequenceMap.begin(); it != sequenceMap.end(); ++it )
	{
		const SequenceMap::key_type &fixes = it->first;
		const Frames &frames = it->second;
		// todo: could be more efficient by writing a custom comparison function that uses indexes 
		//	 into the const Frames vector rather than duplicating the strings and sorting them directly
		Frames sortedFrames = frames;
		std::sort( sortedFrames.begin(), sortedFrames.end() );
		
		/// in diabolical cases the elements of frames may not all have the same padding
		/// so we'll sort them out into padded and unpadded frame sequences here, by creating
		/// a map of padding->list of frames. unpadded things will be considered to have a padding
		/// of 1.
		typedef std::vector< FrameList::Frame > NumericFrames;
		typedef std::map< unsigned int, NumericFrames > PaddingToFramesMap;
		PaddingToFramesMap paddingToFrames;
		for ( Frames::const_iterator fIt = sortedFrames.begin(); fIt != sortedFrames.end(); ++fIt )
		{
			std::string frame = *fIt;
			int sign = 1;

			assert( frame.size() );
			if ( *frame.begin() == '-' )
			{
				frame = frame.substr( 1, frame.size() - 1 );
				sign = -1;
			}
			if ( *frame.begin() == '0' || paddingToFrames.find( frame.size() ) != paddingToFrames.end() )
			{
				paddingToFrames[ frame.size() ].push_back( sign * boost::lexical_cast<FrameList::Frame>( frame ) );
			}
			else
			{
				paddingToFrames[ 1 ].push_back( sign * boost::lexical_cast<FrameList::Frame>( frame ) );
			}
		}

		for ( PaddingToFramesMap::iterator pIt = paddingToFrames.begin(); pIt != paddingToFrames.end(); ++pIt )
		{
			const PaddingToFramesMap::key_type &padding = pIt->first;
			NumericFrames &numericFrames = pIt->second;
			std::sort( numericFrames.begin(), numericFrames.end() );

			FrameListPtr frameList = frameListFromList( numericFrames );

			std::vector< FrameList::Frame > expandedFrameList;
			frameList->asList( expandedFrameList );

			/// remove any sequences with less than the given minimum.
			if ( expandedFrameList.size() >= minSequenceSize )
			{
				std::string frameTemplate;
				for ( PaddingToFramesMap::key_type i = 0; i < padding; i++ )
				{
					frameTemplate += "#";
				}

				sequences.push_back(
					new FileSequence(
						fixes.first + frameTemplate + fixes.second,
						frameList
					)
				);
			}
		}
	}
}

void IECore::findSequences( const std::vector< std::string > &names, std::vector< FileSequencePtr > &sequences )
{
	/// ignore any sequences with less than two files
	findSequences( names, sequences, 2 );
}

void IECore::ls( const std::string &path, std::vector< FileSequencePtr > &sequences, size_t minSequenceSize )
{
	sequences.clear();

	if ( boost::filesystem::is_directory( path ) )
	{
		boost::filesystem::directory_iterator end;
	 	std::vector< std::string > files;
		for ( boost::filesystem::directory_iterator it( path ); it != end; ++it )
		{
			files.push_back( it->path().PATH_TO_STRING );
		}

		findSequences( files, sequences, minSequenceSize );
	}
}

void IECore::ls( const std::string &sequencePath, FileSequencePtr &sequence, size_t minSequenceSize )
{
	sequence = 0;
	boost::smatch matches;
	bool m = boost::regex_match( sequencePath, matches, FileSequence::fileNameValidator() );
	if ( !m )
	{
		return;
	}

	const std::string paddingStr( matches[2].first, matches[2].second );
	const unsigned padding = paddingStr.size();

 	std::vector< std::string > files;

	boost::filesystem::path dir = boost::filesystem::path( sequencePath ).parent_path();

	std::string baseSequencePath = boost::filesystem::path( sequencePath ).PATH_TO_STRING;

	const std::string::size_type first = baseSequencePath.find_first_of( '#' );
	assert( first != std::string::npos );
	const std::string prefix = baseSequencePath.substr( 0, first );

	const std::string::size_type last = baseSequencePath.find_last_of( '#' );
	assert( last != std::string::npos );
	const std::string suffix = baseSequencePath.substr( last + 1, baseSequencePath.size() - last - 1 );

	boost::filesystem::path dirToCheck( dir );
	if ( dirToCheck.string() == "" )
	{
		dirToCheck = ".";
	}

	boost::filesystem::directory_iterator end;

	for ( boost::filesystem::directory_iterator it( dirToCheck ); it != end; ++it )
	{
		const std::string fileName = it->path().PATH_TO_STRING;

		if ( fileName.size() >= std::min( prefix.size(), suffix.size() ) && fileName.substr( 0, prefix.size() ) == prefix && fileName.substr( fileName.size() - suffix.size(), suffix.size() ) == suffix )
		{
			files.push_back( ( dir / boost::filesystem::path( fileName ) ).string() );
		}
	}

	std::vector< FileSequencePtr > sequences;
	findSequences( files, sequences, minSequenceSize );

	for ( std::vector< FileSequencePtr >::iterator it = sequences.begin(); it != sequences.end() ; ++it )
	{
		if ( (*it)->getPadding() == padding )
		{
			sequence = *it;
			return;
		}
		// Also accept frame ranges with number of digits greater than the padding.
		if ( (*it)->getPadding() == 1 )
		{
			FrameList::Frame startFrame, endFrame;
			
			if( FrameRange *fr = runTimeCast<FrameRange>( (*it)->getFrameList().get() ) )
			{
				// Can do a quick validation if the range is a FrameRange, by looking at the start and end only
				startFrame = fr->getStart();
				endFrame = fr->getEnd();
			}
			else
			{
				// Otherwise, have to get the full list
				std::vector<FrameList::Frame> frames;
				(*it)->getFrameList()->asList(frames);
				// I know that there will be at least one frame, since the file sequence was found
				startFrame = frames[ 0 ];
				endFrame = frames[ 0 ];
				for ( unsigned i=0; i < frames.size(); i++ )
				{
					if ( frames[ i ] < startFrame )
					{
						startFrame = frames[ i ];
					}
					else if ( frames[ i ] > endFrame )
					{
						endFrame = frames[ i ];
					}
				}
			}
			
			// if goes through 0 then this is a no.
			if ( !(startFrame < 0 && endFrame > 0) )
			{
				unsigned int startDigits = 1;
				if ( startFrame > 0 )
					startDigits = static_cast<unsigned int>(floor(log10( static_cast<double>(abs(startFrame)) ))) + 1;

				if ( startDigits >= padding )
				{
					sequence = *it;
					sequence->setPadding( padding );
					return;
				}
			}
		}
	}
}

FrameListPtr IECore::frameListFromList( const std::vector< FrameList::Frame > &frames )
{
	if ( frames.size() == 0 )
	{
		return new EmptyFrameList();
	}
	else if ( frames.size() == 1 )
	{
		return new FrameRange( frames[0], frames[0] );
	}

	std::vector< FrameListPtr > frameLists;

	FrameList::Frame rangeStart = 0;
	FrameList::Frame rangeEnd = 1;
	FrameList::Frame rangeStep = frames[ rangeEnd ] - frames[ rangeStart ];
	assert( rangeStep > 0 );

	while ( rangeEnd <= (FrameList::Frame)frames.size() )
	{
		if ( rangeEnd == (FrameList::Frame)frames.size() || frames[ rangeEnd ] - frames[ rangeEnd -1 ] != rangeStep )
		{
			/// we've come to the end of a run
			if ( rangeEnd - 1 == rangeStart )
			{
				frameLists.push_back( new FrameRange( frames[ rangeStart ], frames[ rangeStart ] ) );
			}
			else
			{
				if ( rangeStep > 0)
				{
					frameLists.push_back( new FrameRange( frames[ rangeStart ], frames[ rangeEnd -1 ], rangeStep ) );
				}
				else
				{
					frameLists.push_back( new ReversedFrameList( new FrameRange(frames[ rangeEnd -1 ], frames[ rangeStart ], -rangeStep ) ));
				}
			}

			rangeStart = rangeEnd;
			rangeEnd = rangeStart + 1;

			if ( rangeEnd < (FrameList::Frame)frames.size() )
			{
				rangeStep = frames[ rangeEnd ] - frames[ rangeStart ];
			}
		}
		else
		{
			rangeEnd ++;
		}
	}

	if ( frameLists.size() == 1 )
	{
		return frameLists[0];
	}
	else
	{
		return new CompoundFrameList( frameLists );
	}

}

File: python/IECore/FileSequenceFunctions.py
##########################################################################
#
#  Copyright (c) 2007-2009, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

## \file FileSequenceFunctions.py
# Defines functions useful for creating and manipulating FileSequence objects.
#
# \ingroup python

import os
import re
import glob
import shutil
import os.path
from FileSequence import FileSequence
import _IECore as IECore

## Returns a list of FileSequence objects representing all the sequences in names.
# names is just a list of arbitrary strings, which may or may not represent files
# on disk.
def findSequences( names ) :

	# this matches names of the form $prefix$frameNumber$suffix
	# placing each of those in a group of the resulting match.
	# both $prefix and $suffix may be the empty string and $frameNumber
	# may be preceded by a minus sign.
	matchExpression = re.compile( "^([^#]*?)(-{,1}[0-9]+)([^0-9#]*)$" )
	# build a mapping from ($prefix, $suffix) to a list of $frameNumbers
	sequenceMap = {}
	for f in names :
	
		match = matchExpression.match( f )
		if match :
		
			fixes = match.group( 1 ), match.group( 3 )
			frames = sequenceMap.setdefault( fixes, [] )
			frames.append( match.group( 2 ) )
	
	# build a list of FileSequence objects to return
	result = []	
	for fixes, frames in sequenceMap.items() :
	
		# in diabolical cases the elements of frames may not all have the same padding
		# so we'll sort them out into padded and unpadded frame sequences here, by creating
		# a map of padding->list of frames. unpadded things will be considered to have a padding
		# of 1.
		frames.sort()
		paddingToFrames = {}
		for frame in frames :
			sign = 1
			if frame[0]=="-" :
				frame = frame[1:]
				sign = -1
			if frame[0]=="0" or len(frame) in paddingToFrames :
				numericFrames = paddingToFrames.setdefault( len(frame), [] )
			else :
				numericFrames = paddingToFrames.setdefault( 1, [] )
			numericFrames.append( sign * int(frame) )
		
		for padding, numericFrames in paddingToFrames.items() :
			
			numericFrames.sort()	
			frameList = frameListFromList( numericFrames )
			result.append( FileSequence( fixes[0] + "".ljust( padding, "#" ) + fixes[1], frameList ) )
	
	# remove any sequences with less than two files
	result = [ x for x in result if len( x.frameList.asList() ) > 1 ]
		
	return result

## If path is a directory then returns all sequences
# residing in that directory in the form of a list of FileSequences, returning
# the empty list if none are found. If path is a sequence style filename with a
# single sequence of at least one embedded # then tries to find a sequence
# matching that specification, returning None if one isn't found.
def ls( path ) :
	
	r = FileSequence.fileNameValidator()
	if r.match( path ) :
	
		padding = len( r.match( path ).group( 2 ) )
		globPath = re.sub( "#+", "*", path )
				
		files = glob.glob( globPath )
				
		sequences = findSequences( files )
		for sequence in sequences :
			if sequence.getPadding()==padding :
				return sequence
				
		return None
				
	elif os.path.isdir( path ) :
		
		allFiles = os.listdir( path )
		return findSequences( allFiles )			

## Moves the set of files specified by sequence1 to the set of files
# specified by sequence2, where sequence1 and sequence2 are
# FileSequence objects of equal length. This function is safe even if the
# files specified by each sequence overlap.				
def mv( sequence1, sequence2 ) :

	if __sequencesClash( sequence1, sequence2 ) :
		sTmp = sequence1.copy()
		sTmp.setPrefix( os.path.join( os.path.dirname( sTmp.getPrefix() ), __tmpPrefix() ) )
		for src, dst in sequence1.mapTo( sTmp, True ) :
			shutil.move( src, dst )
		for src, dst in sTmp.mapTo( sequence2, True ) :
			shutil.move( src, dst )
	else :
		for src, dst in sequence1.mapTo( sequence2, True ) :
			shutil.move( src, dst )

## Copies the set of files specified by sequence1 to the set of files
# specified by sequence2, where sequence1 and sequence2 are
# FileSequence objects of equal length. This function is safe even if the
# files specified by each sequence overlap.	
def cp( sequence1, sequence2 ) :
	
	if __sequencesClash( sequence1, sequence2 ) :
		raise RuntimeError( "Attempt to copy sequences with common filenames." )
		
	for src, dst in sequence1.mapTo( sequence2, True ) :
		shutil.copy( src, dst )

## Removes all the files specified by the sequence.
def rm( sequence ) :
	
	for f in sequence.fileNames() :
	
		os.remove( f )

## Concetenates all the files specified by the sequence to stdout
# \todo Allow destination file to be specified
def cat( sequence ) :	

	for f in sequence.fileNames() :
	
		ret = os.system( 'cat "%s"' % ( f ) )	
		
		

## Returns a FrameList instance that "best" represents the specified list of integer
# frame numbers. This function attempts to be intelligent and uses a CompoundFrameList
# of FrameRange objects to represent the specified frames compactly.
def frameListFromList( frames ) :

	if len( frames )==0 :
		return IECore.EmptyFrameList()
		
	if len( frames )==1 :
		return IECore.FrameRange( frames[0], frames[0] )
		
	frameLists = []
	
	rangeStart = 0
	rangeEnd = 1
	rangeStep = frames[rangeEnd] - frames[rangeStart]
	assert( rangeStep > 0 )
	
	while rangeEnd<=len( frames ) :
		
		if rangeEnd==len( frames ) or frames[rangeEnd] - frames[rangeEnd-1] != rangeStep :
			# we've come to the end of a run
			if rangeEnd-1==rangeStart :
				frameLists.append( IECore.FrameRange( frames[rangeStart], frames[rangeStart] ) )
			else :
				frameLists.append( IECore.FrameRange( frames[rangeStart], frames[rangeEnd-1], rangeStep ) )
			rangeStart = rangeEnd
			rangeEnd = rangeStart + 1
				
			if rangeEnd<len( frames ) :
				rangeStep = frames[rangeEnd] - frames[rangeStart]
		else :
			# continue the current run
			rangeEnd += 1

	if len( frameLists )==1 :
		return frameLists[0]
	else :
		return IECore.CompoundFrameList( frameLists )

# private utility functions

def __sequencesClash( sequence1, sequence2 ) :

	s = set()
	for f in sequence1.fileNames() :
		s.add( f )
		
	for f in sequence2.fileNames() :
		if f in s :
			return True

	return False
	
def __tmpPrefix() :

	"""Returns a hopefully unique string suitable for use as the temporary
	sequence prefix when moving a sequence."""
	
	import hashlib
	import platform
	import time
	
	h = hashlib.md5()
	h.update( platform.node() ) # computer name
	h.update( str( os.getpid() ) )
	h.update( str( time.time() ) )
	return "ieSequenceTmp" + h.hexdigest() + "."

__all__ = [ "findSequences", "ls", "mv", "cp", "rm", "cat", "frameListFromList" ]
